CorpusCallosum/segmentation/segmentation_inference.py

```diff
@@ -0,0 +1,279 @@
+        outputs_avg = torch.stack([post_trans(i) for i in outputs_avg])
+        
+        pad_left, pad_right, pad_top, pad_bottom = to_pad
+        # Pad back to original size
+        outputs = np.pad(outputs, ((0,0), (0,0), (pad_left.item(), pad_right.item()), 
+                                   (pad_top.item(), pad_bottom.item())), mode='constant', constant_values=0)
+        outputs_avg = np.pad(outputs_avg, ((0,0), (0,0), (pad_left.item(), pad_right.item()), 
+                                           (pad_top.item(), pad_bottom.item())), mode='constant', constant_values=0)
+        outputs_soft = np.pad(outputs_soft, ((0,0), (0,0), (pad_left.item(), pad_right.item()), 
+                                             (pad_top.item(), pad_bottom.item())), mode='constant', constant_values=0)
+    
+    # restore original shape
+    if orig_shape[-2:] != outputs.shape[-2:]:
+        new_outputs = np.zeros((outputs.shape[0], outputs.shape[1], orig_shape[-2], orig_shape[-1]))
+        new_outputs[:,:,:256,:256] = outputs
+        outputs = new_outputs
+        
+        new_outputs_avg = np.zeros((outputs_avg.shape[0], outputs_avg.shape[1], orig_shape[-2], orig_shape[-1]))
+        new_outputs_avg[:,:,:256,:256] = outputs_avg
+        outputs_avg = new_outputs_avg
+
+        new_outputs_soft = np.zeros((outputs_soft.shape[0], outputs_soft.shape[1], 
+                                     orig_shape[-2], orig_shape[-1]), dtype=np.float32)
+        new_outputs_soft[:,:,:256,:256] = outputs_soft
+        outputs_soft = new_outputs_soft
+
+    return (
+        outputs.transpose(0,2,3,1),
+        inputs.cpu().numpy().transpose(0,2,3,1),
+        outputs_avg.transpose(0,2,3,1),
+        outputs_soft.transpose(0,2,3,1),
+    )
+
+# TODO: load validation data and run inference on it to confirm correct processing
+
+
+def load_validation_data(path):
+    import pandas as pd
+    data = pd.read_csv(path, index_col=0, header=None)
+    data.columns = ["image", "label", "AC_center_x", "AC_center_y", "AC_center_z", 
+                    "PC_center_x", "PC_center_y", "PC_center_z"]
+    
+    ac_centers = data[["AC_center_x", "AC_center_y", "AC_center_z"]].values
+    pc_centers = data[["PC_center_x", "PC_center_y", "PC_center_z"]].values
+    images = data["image"].values
+    labels = data["label"].values
+    subj_ids = data.index.values.tolist()
+
+    label_widths = []
+    for label_path in data['label']:
+        label_img =nib.load(label_path)
+        
+        if label_img.shape[0] > 100:
+            # check which slices have non-zero values
+            label = label_img.get_fdata()
+            non_zero_slices = np.any(label > 0, axis=(1,2))
+            first_nonzero = np.argmax(non_zero_slices)
+            last_nonzero = len(non_zero_slices) - np.argmax(non_zero_slices[::-1])
+            label_widths.append(last_nonzero - first_nonzero)
+        else:
+            label_widths.append(label_img.shape[0])
+        
+    
```

Comment:
This can probably significantly benefit from parallelization.
```suggestion
    def _load(label_path: str | Path) -> int:
        label_img = nib.load(label_path)
        
        if label_img.shape[0] > 100:
            # check which slices have non-zero values
            label_data = np.asarray(label_img.dataobj)
            non_zero_slices = np.any(label_data > 0, axis=(1,2))
            first_nonzero = np.argmax(non_zero_slices)
            last_nonzero = len(non_zero_slices) - np.argmax(non_zero_slices[::-1])
            return last_nonzero - first_nonzero
        else:
            return label_img.shape[0]
    label_widths = ThreadPoolExecutor().map(_load, data["label"])    
```